File: data-structures/stack/tests/stack_test.go
package tests

import (
	"reflect"
	"testing"

	"dsa/data-structures/stack/implementations"
)

func createStackReferenceSequence() []int {
	values := make([]int, 0)

	for index := 1; index <= 50; index++ {
		values = append(values, index)
		if index%4 == 0 {
			values = values[:len(values)-1]
		}
		if index%7 == 0 {
			values = append(values, index*10)
		}
	}

	return values
}

func TestStackSimpleNewStackStartsEmpty(t *testing.T) {
	stack := implementations.NewMyStack[int](2)

	if got := stack.Length(); got != 0 {
		t.Fatalf("expected length 0, got %d", got)
	}

	if got := stack.Capacity(); got != 2 {
		t.Fatalf("expected capacity 2, got %d", got)
	}

	if !stack.IsEmpty() {
		t.Fatal("expected stack to be empty")
	}

	if got := stack.ToSlice(); len(got) != 0 {
		t.Fatalf("expected empty slice, got %v", got)
	}
}

func TestStackSimplePushAndPeekExposeCurrentTop(t *testing.T) {
	stack := implementations.NewMyStack[int](2)

	stack.Push(10)
	stack.Push(20)
	stack.Push(30)

	if got := stack.Length(); got != 3 {
		t.Fatalf("expected length 3, got %d", got)
	}

	if got := stack.Peek(); got != 30 {
		t.Fatalf("expected top value 30, got %d", got)
	}

	if got := stack.ToSlice(); !reflect.DeepEqual(got, []int{10, 20, 30}) {
		t.Fatalf("expected [10 20 30], got %v", got)
	}
}

func TestStackSimplePopReturnsItemsInLIFOOrder(t *testing.T) {
	stack := implementations.NewMyStack[string](2)

	stack.Push("a")
	stack.Push("b")
	stack.Push("c")

	if got := stack.Pop(); got != "c" {
		t.Fatalf("expected c, got %s", got)
	}

	if got := stack.Pop(); got != "b" {
		t.Fatalf("expected b, got %s", got)
	}

	if got := stack.Peek(); got != "a" {
		t.Fatalf("expected a, got %s", got)
	}
}

func TestStackSimpleClearResetsTheStack(t *testing.T) {
	stack := implementations.NewMyStack[int](2)

	stack.Push(1)
	stack.Push(2)
	stack.Clear()

	if got := stack.Length(); got != 0 {
		t.Fatalf("expected length 0, got %d", got)
	}

	if !stack.IsEmpty() {
		t.Fatal("expected stack to be empty")
	}
}

func TestStackMildlyAbsurdGrowthBeyondInitialCapacity(t *testing.T) {
	stack := implementations.NewMyStack[int](1)

	for value := 0; value < 25; value++ {
		stack.Push(value)
	}

	if got := stack.Length(); got != 25 {
		t.Fatalf("expected length 25, got %d", got)
	}

	if got := stack.Capacity(); got < 25 {
		t.Fatalf("expected capacity to grow beyond 25, got %d", got)
	}
}

func TestStackMildlyAbsurdAlternatingPushAndPopLeavesRightSurvivors(t *testing.T) {
	stack := implementations.NewMyStack[int](2)

	stack.Push(1)
	stack.Push(2)
	stack.Pop()
	stack.Push(3)
	stack.Push(4)
	stack.Pop()
	stack.Push(5)

	if got := stack.ToSlice(); !reflect.DeepEqual(got, []int{1, 3, 5}) {
		t.Fatalf("expected [1 3 5], got %v", got)
	}
}

func TestStackAbsurdDeterministicOperationSequenceMatchesReferenceModel(t *testing.T) {
	stack := implementations.NewMyStack[int](2)
	reference := createStackReferenceSequence()

	for _, value := range reference {
		stack.Push(value)
	}

	if got := stack.ToSlice(); !reflect.DeepEqual(got, reference) {
		t.Fatalf("expected %v, got %v", reference, got)
	}
}

File: data-structures/stack/tests/stack.test.ts
import assert from "node:assert/strict";
import { MyStack } from "../implementations/stack";

function runTest(name: string, execute: () => void): void {
  try {
    execute();
    console.log(`PASS ${name}`);
  } catch (error) {
    console.error(`FAIL ${name}`);
    throw error;
  }
}

function createReferenceSequence(): number[] {
  const values: number[] = [];

  for (let index = 1; index <= 50; index += 1) {
    values.push(index);
    if (index % 4 === 0) {
      values.pop();
    }
    if (index % 7 === 0) {
      values.push(index * 10);
    }
  }

  return values;
}

runTest("simple: new stack starts empty", () => {
  const stack = new MyStack<number>(2);

  assert.equal(stack.length(), 0);
  assert.equal(stack.capacity(), 2);
  assert.equal(stack.isEmpty(), true);
  assert.equal(stack.peek(), undefined);
  assert.deepEqual(stack.toArray(), []);
});

runTest("simple: push and peek expose the current top", () => {
  const stack = new MyStack<number>(2);

  stack.push(10);
  stack.push(20);
  stack.push(30);

  assert.equal(stack.length(), 3);
  assert.equal(stack.peek(), 30);
  assert.deepEqual(stack.toArray(), [10, 20, 30]);
});

runTest("simple: pop returns items in LIFO order", () => {
  const stack = new MyStack<string>(2);

  stack.push("a");
  stack.push("b");
  stack.push("c");

  assert.equal(stack.pop(), "c");
  assert.equal(stack.pop(), "b");
  assert.equal(stack.peek(), "a");
  assert.deepEqual(stack.toArray(), ["a"]);
});

runTest("simple: clear resets the stack", () => {
  const stack = new MyStack<number>(2);

  stack.push(1);
  stack.push(2);
  stack.clear();

  assert.equal(stack.length(), 0);
  assert.equal(stack.isEmpty(), true);
  assert.equal(stack.pop(), undefined);
});

runTest("mildly absurd: stack grows beyond initial capacity", () => {
  const stack = new MyStack<number>(1);

  for (let value = 0; value < 25; value += 1) {
    stack.push(value);
  }

  assert.equal(stack.length(), 25);
  assert.ok(stack.capacity() >= 25);
  assert.equal(stack.peek(), 24);
});

runTest("mildly absurd: alternating push and pop leaves the right survivors", () => {
  const stack = new MyStack<number>(2);

  stack.push(1);
  stack.push(2);
  stack.pop();
  stack.push(3);
  stack.push(4);
  stack.pop();
  stack.push(5);

  assert.deepEqual(stack.toArray(), [1, 3, 5]);
  assert.equal(stack.peek(), 5);
});

runTest("absurd: deterministic operation sequence matches a reference model", () => {
  const stack = new MyStack<number>(2);
  const reference = createReferenceSequence();

  for (const value of reference) {
    stack.push(value);
  }

  assert.deepEqual(stack.toArray(), reference);
  assert.equal(stack.length(), reference.length);
  assert.equal(stack.peek(), reference[reference.length - 1]);
});
